### PR TITLE
Refactor `tensor_search::search` to use shared functions from `tensor_search::bulk_search`

### PR DESCRIPTION
**Pull request by Jeadie**
## Changeset
 - Bulk search now supports
   - Context vectors
   - Score modifiers
 - Standard search is now refactored to use modular functions used in bulk search.

### CODE REVIEW

File: src/marqo/tensor_search/tensor_search.py [hunk 1]
```diff
@@ -1341,65 +1337,56 @@ def construct_vector_input_batches(query: Union[str, Dict], index_info) -> Tuple
...
         searchable_attributes_as_vectors = {
             utils.generate_vector_name(field_name=attribute) for attribute in searchable_attributes
         }
         # discard searchable attributes that aren't found in the cache:
-        return list(searchable_attributes_as_vectors.intersection(
+        properties_to_search = list(searchable_attributes_as_vectors.intersection(
             index_info.get_vector_properties().keys()
         ))
+    
+    # Validation for offset (pagination is single field) if offset not provided, validation is not run.
```
Comment (Jeadie) [inline src/marqo/tensor_search/tensor_search.py:1314]: Moving into here so it can be used by both search/ bulk_search

File: src/marqo/tensor_search/tensor_search.py [hunk 2]
```diff
@@ -1341,65 +1337,56 @@ def construct_vector_input_batches(query: Union[str, Dict], index_info) -> Tuple
...
+                                            list(properties_to_search)]
+        raise errors.InvalidArgError(
+            f"Pagination (offset > 0) is only supported for single field searches!"
+            f" Your search currently has {len(properties_to_search)} vectorisable fields: {human_readable_vector_properties}"
+        )
+    return properties_to_search
 
 
-def construct_msearch_body_elements(vector_properties_to_search: List[str], offset: int, filter_string: str, index_info: IndexInfo, result_count: int, query_vector: List[float], attributes_to_retrieve: List[str], index_name: str, contextualised_filter: str) -> List[Dict[str, Any]]:
+def construct_msearch_body_elements(searchableAttributes: List[str], offset: int, filter_string: str, index_info: IndexInfo, result_count: int, query_vector: List[float], attributes_to_retrieve: List[str], index_name: str, score_modifiers: Optional[Dict[str, Any]] = None) -> List[Dict[str, Any]]:
```
Comment (Jeadie) [inline src/marqo/tensor_search/tensor_search.py]: This change lets:
1. search use `construct_msearch_body_elements`
2. Bulk search to use score modifiers

File: src/marqo/tensor_search/tensor_search.py [hunk 3]
```diff
@@ -1419,14 +1406,17 @@ def bulk_msearch(config: Config, body: List[Dict]) -> List[Dict]:
     except KeyError as e:
         # KeyError indicates we have received a non-successful result
         try:
-            if "index.max_result_window" in response["responses"][0]["error"]["root_cause"][0]["reason"]:
```
Comment (Jeadie) [inline src/marqo/tensor_search/tensor_search.py:1384]: This just makes it cleaner, and adds 
`elif  root_cause_type == 'query_shard_exception' and root_cause_reason.startswith("Failed to parse query")` to replicate what is in search.

File: src/marqo/tensor_search/tensor_search.py [hunk 4]
```diff
@@ -1595,7 +1585,19 @@ def get_query_vectors_from_jobs(
             ]
             # TODO how doe we ensure order?
             weighted_vectors = [np.asarray(vec) * weight for vec, weight, content in vectorised_ordered_queries]
-            merged_vector = np.mean(weighted_vectors, axis=0)
+            
+            custom_tensors = q.get_context_tensor() 
+            if custom_tensors is not None:
```
Comment (Jeadie) [inline src/marqo/tensor_search/tensor_search.py]: custom tensor support for bulk_search (and search to use this method).

File: src/marqo/tensor_search/tensor_search.py [hunk 5]
```diff
@@ -1647,6 +1649,25 @@ def create_empty_query_response(queries: List[BulkSearchQueryEntity]) -> List[Di
         )
     )
 
+def run_vectorise_pipeline(config: Config, queries: List[BulkSearchQueryEntity], selected_device: Union[Device, str]) -> Dict[Qidx, List[float]]:
```
Comment (Jeadie) [inline src/marqo/tensor_search/tensor_search.py:1628]: This abstracts the query -> vector pipeline so we can use it for (bulk) search

File: tests/tensor_search/test_lexical_search.py [hunk 6]
```diff
@@ -128,36 +127,7 @@ def test_lexical_search_multiple_searchable_attribs(self):
             docs=[d3, d2])
         res = tensor_search._lexical_search(
             config=self.config, index_name=self.index_name_1, text="Marqo field",
-            return_doc_ids=True, searchable_attributes=["field lambda", "FIELD omega"])
-        assert len(res["hits"]) == 3
-        assert self.strip_marqo_fields(res["hits"][0]) == d1
-        assert self.strip_marqo_fields(res["hits"][1]) == d3
-        assert self.strip_marqo_fields(res["hits"][2]) == d4
-
-    def test_lexical_search_multiple_searchable_attribs_no_returned_ids(self):
```
Comment (Jeadie) [inline tests/tensor_search/test_lexical_search.py:143]: This is no longer supported.

File: tests/tensor_search/test_search.py [hunk 7]
```diff
@@ -132,40 +131,9 @@ def test_vector_search_long_query_string(self):
                 {"_id": "abc12334", "Title": "Grandma Jo's family recipe. ",
                  "Steps": "1. Cook meat. 2: Dice Onions. 3: Serve."},
             ], auto_refresh=True)
-        search_res = tensor_search._vector_text_search(
-            config=self.config, index_name=self.index_name_1, query=query_text,
-            return_doc_ids=True
-        )
-
-    def test_vector_search_all_highlights(self):
```
Comment (Jeadie) [inline tests/tensor_search/test_search.py:141]: because we are always in simplified format, these tests don't apply.

**Comment by Jeadie** [comment]
https://github.com/marqo-ai/marqo/actions/runs/4932152115

**Comment by Jeadie** [comment]
https://github.com/marqo-ai/marqo/actions/runs/4954372137

File: src/marqo/tensor_search/models/api_models.py [hunk 8]
```diff
@@ -27,8 +32,8 @@ class SearchQuery(BaseMarqoModel):
     attributesToRetrieve: Union[None, List[str]] = None
     boost: Optional[Dict] = None
     image_download_headers: Optional[Dict] = None
-    context: Optional[Dict] = None
-    scoreModifiers: Optional[Dict] = None
+    context: Optional[SearchContext] = None
+    scoreModifiers: Optional[ScoreModifier] = None,
```
Comment (pandu-k) [inline src/marqo/tensor_search/models/api_models.py]: unnecessary comma

File: src/marqo/tensor_search/models/score_modifiers_object.py [hunk 9]
```diff
@@ -1,65 +1,95 @@
...
+
+
+class ScoreModifierOperator(BaseModel):
+    field_name: str
+    weight: float = 1
+
+    class Config:
+        extra: str = "forbid"
+
+    @validator('field_name')
```
Comment (pandu-k) [inline src/marqo/tensor_search/models/score_modifiers_object.py:26]: We may need more field validation. There is a field validation function in validation.py that  you can use

File: src/marqo/tensor_search/models/score_modifiers_object.py [hunk 10]
```diff
@@ -1,65 +1,95 @@
...
+    def name_not_id(cls, v):
+        if v == "_id":
+            raise InvalidArgError("_id is not allowed as a field_name")
+        return v
+
+    def to_painless_script(self, operation: Literal["multiply_score_by", "add_to_score"]) -> str:
```
Comment (pandu-k) [inline src/marqo/tensor_search/models/score_modifiers_object.py:32]: I prefer minimal logic to be within object methods (esp. if it mutates object state). If possible can you use module levels functions instead?

File: src/marqo/tensor_search/models/score_modifiers_object.py [hunk 11]
```diff
@@ -1,65 +1,95 @@
...
+            f"Please revise your score_modifiers based on the provided error."
+            f"\n Check `https://docs.marqo.ai/0.0.17/API-Reference/search/#score-modifiers` for more info."
+        )
+
+
+class ScoreModifierOperator(BaseModel):
+    field_name: str
+    weight: float = 1
+
+    class Config:
```
Comment (pandu-k) [inline src/marqo/tensor_search/models/score_modifiers_object.py:22]: Can we make it, and any pydantic object that we can, immutable? via `allow_mutation = False`

File: src/marqo/tensor_search/models/score_modifiers_object.py
Comment (pandu-k) [inline src/marqo/tensor_search/models/score_modifiers_object.py:1]: 2 main points:
1. Make pydantic models immutable in 99.9%, unless there's a good reason not too
2. Lots of unit tests, including edge cases, for all logic

File: src/marqo/tensor_search/tensor_search.py
Comment (pandu-k) [inline src/marqo/tensor_search/tensor_search.py:1]: Possibly move all the search functions to a new file/dir: `tensor_search/search.py` or `tensor_search/search/search.py`

Only if this doesn't add too much extra work to the refactor. Otherwise we can do this refactor some other time

File: src/marqo/tensor_search/models/score_modifiers_object.py
Comment (pandu-k) [reply src/marqo/tensor_search/models/score_modifiers_object.py:1]: Including object methods. Even super trivial unit tests help Marqo more stable

File: src/marqo/tensor_search/models/score_modifiers_object.py (hunk 10 above)
Comment (Jeadie) [reply src/marqo/tensor_search/models/score_modifiers_object.py:32]: This doesn't mutate state and is in essence just a different view on the underlying data.

File: src/marqo/tensor_search/models/score_modifiers_object.py (hunk 9 above)
Comment (Jeadie) [reply src/marqo/tensor_search/models/score_modifiers_object.py:26]: This is true, although I think is additional validation (because ScoreModifierOperator.fields are validated at field creation time and when ScoreModifierOperator looks for the field name.

File: src/marqo/tensor_search/tensor_search.py
Comment (Jeadie) [reply src/marqo/tensor_search/tensor_search.py:1]: I don't want to move much of this until the `add_documents`. Then I will know if/what is search-only and add_documents-only

File: src/marqo/tensor_search/tensor_search.py [hunk 12]
```diff
@@ -1264,9 +1261,8 @@ def _lexical_search(
     res_list = []
     for doc in search_res['hits']['hits']:
         just_doc = _clean_doc(doc["_source"].copy()) if "_source" in doc else dict()
-        if return_doc_ids:
-            just_doc["_id"] = doc["_id"]
-            just_doc["_score"] = doc["_score"]
+        just_doc["_id"] = doc["_id"]
```
Comment (vicilliar) [inline src/marqo/tensor_search/tensor_search.py:1264]: are we sure there's no instance wherein we don't want to return doc ids?

File: src/marqo/tensor_search/tensor_search.py [hunk 13]
```diff
@@ -1131,7 +1130,7 @@ def search(config: Config, index_name: str, text: Union[str, dict],
                                          model_name=reranker,
                                          device=config.indexing_device if device is None else device,
                                          searchable_attributes=searchable_attributes,
-                                         num_highlights=1 if simplified_format else num_highlights)
```
Comment (vicilliar) [inline src/marqo/tensor_search/tensor_search.py:1134]: also is there no case where we'd want more than 1 highlight?

**Review by vicilliar — COMMENTED** [review]
looks good! just added some questions

File: src/marqo/tensor_search/models/score_modifiers_object.py [hunk 14]
```diff
@@ -1,65 +1,95 @@
...
+                if (doc['__chunks.{self.field_name}'].size() > 0 &&
+                    (doc['__chunks.{self.field_name}'].value instanceof java.lang.Number)) {{
+                    additive = additive + doc['__chunks.{self.field_name}'].value * {self.weight};
+                }}
+            }}
+            """
+
+        else:
+            raise ValueError(
+                f"operation must be either 'multiplicative' or 'additive', {operation} provided."
```
Comment (vicilliar) [inline src/marqo/tensor_search/models/score_modifiers_object.py]: shouldn't this message say ""multiply_score_by" or "add_to_score"?

File: src/marqo/tensor_search/tensor_search.py (hunk 13 above)
Comment (Jeadie) [reply src/marqo/tensor_search/tensor_search.py:1134]: Theres's not way to set it.

File: src/marqo/tensor_search/tensor_search.py (hunk 12 above)
Comment (Jeadie) [reply src/marqo/tensor_search/tensor_search.py:1264]: as above, not in API

File: src/marqo/tensor_search/models/score_modifiers_object.py (hunk 10 above)
Comment (pandu-k) [reply src/marqo/tensor_search/models/score_modifiers_object.py:32]: All good, as long as there are sufficient tests

File: src/marqo/tensor_search/tensor_search.py
Comment (pandu-k) [reply src/marqo/tensor_search/tensor_search.py:1]: Ok, that makes sense

File: src/marqo/tensor_search/tensor_search.py (hunk 1 above)
Comment (pandu-k) [reply src/marqo/tensor_search/tensor_search.py:1314]: nit: perhaps be moved to a separate validation function (may be out-of-scope for this refactor)

File: src/marqo/tensor_search/tensor_search.py (hunk 3 above)
Comment (pandu-k) [reply src/marqo/tensor_search/tensor_search.py:1384]: good to clean this up 👍

File: src/marqo/tensor_search/tensor_search.py (hunk 4 above)
Comment (pandu-k) [reply src/marqo/tensor_search/tensor_search.py]: should we call this var `context_tensors` to be consistent?

File: src/marqo/tensor_search/tensor_search.py [hunk 15]
```diff
@@ -1559,7 +1560,19 @@ def get_query_vectors_from_jobs(
...
+            
+            custom_tensors = q.get_context_tensor() 
+            if custom_tensors is not None:
+                weighted_vectors += [np.asarray(v.vector) * v.weight for v in custom_tensors]
+            
+            try:
+                merged_vector = np.mean(weighted_vectors, axis=0)
+            except ValueError as e:
+                raise errors.InvalidArgError(f"The provided vectors are not in the same dimension of the index."
+                                             f"This causes the error when we do `numpy.mean()` over all the vectors.\n"
```
Comment (pandu-k) [inline src/marqo/tensor_search/tensor_search.py:1573]: let's leave out `numpy` as this is a Marqo implementation detail (just `mean` is fine).

File: src/marqo/tensor_search/tensor_search.py (hunk 1 above)
Comment (Jeadie) [reply src/marqo/tensor_search/tensor_search.py:1314]: agree, but this is just a move from within the script.

File: src/marqo/tensor_search/tensor_search.py (hunk 15 above)
Comment (Jeadie) [reply src/marqo/tensor_search/tensor_search.py:1573]: This is a copy paste. I want to minimise refactors during a code reshuffle.

**Review by pandu-k — COMMENTED** [review]
tests are looking good

File: src/marqo/tensor_search/tensor_search.py [hunk 16]
```diff
@@ -1628,9 +1629,9 @@ def run_vectorise_pipeline(config: Config, queries: List[BulkSearchQueryEntity],
     """Run the query vectorisation process"""
     # 1. Pre-process inputs ready for s2_inference.vectorise
     # we can still use qidx_to_job. But the jobs structure may need to be different
-    vector_jobs_tuple: Tuple[Dict[Qidx, List[VectorisedJobPointer]], Dict[JHash, VectorisedJobs]] = (
-        create_vector_jobs(queries, config, selected_device)
-    )
+    vector_jobs_tuple: Tuple[Dict[Qidx, List[VectorisedJobPointer]], Dict[JHash, VectorisedJobs]] = create_vector_jobs(queries, config, selected_device)
+
+    print(vector_jobs_tuple, create_vector_jobs)
```
Comment (pandu-k) [inline src/marqo/tensor_search/tensor_search.py]: rm this print statement

File: tests/tensor_search/test_bulk_search.py [hunk 17]
```diff
@@ -15,25 +15,215 @@
...
+            789: VectorisedJobs(model_name="model_name", model_auth=None, model_properties={}, content_type="text", content=["red herring"], device="cpu",normalize_embeddings=False),
+        }
+        self.qidx_to_job: Dict[Qidx, List[VectorisedJobPointer]] = {
+            0: [VectorisedJobPointer(job_hash=123, start_idx=0, end_idx=1)],
+            1: [VectorisedJobPointer(job_hash=456, start_idx=0, end_idx=1)]
+        }
+        self.job_to_vectors: Dict[JHash, Dict[str, List[float]]] = {
+            123: {"a test query":  [0.5, 0.5]},
+            456: {"another": [0.5, 0.6]},
+            789: {"red herring": [0.6, 0.5]}
```
Comment (pandu-k) [inline tests/tensor_search/test_bulk_search.py:61]: What about a multimodal combination field?

File: src/marqo/tensor_search/tensor_search.py (hunk 16 above)
Comment (Jeadie) [reply src/marqo/tensor_search/tensor_search.py]: ahhh, I ctrl-F ed for prints in my git diff :(

**Review by pandu-k — COMMENTED** [review]
cheers

**Review by pandu-k — APPROVED** [review]
